api/src/main/java/io/cloudevents/message/impl/BaseGenericBinaryMessageImpl.java

```diff
@@ -0,0 +1,71 @@
+package io.cloudevents.message.impl;
+
+import io.cloudevents.SpecVersion;
+import io.cloudevents.message.BinaryMessageVisitor;
+import io.cloudevents.message.BinaryMessageVisitorFactory;
+import io.cloudevents.message.MessageVisitException;
+
+import java.util.Objects;
+import java.util.function.BiConsumer;
+
+/**
+ * This class implements a BinaryMessage, providing commong logic to most protocol bindings
+ * which supports both Binary and Structured mode.
+ * Content-type is handled separately using a key not prefixed with CloudEvents header prefix.
+ *
+ * @param <HV> Header value type
+ */
+public abstract class BaseGenericBinaryMessageImpl<HV> extends BaseBinaryMessage {
+
+    private final SpecVersion version;
+    private final byte[] body;
+
+    protected BaseGenericBinaryMessageImpl(SpecVersion version, byte[] body) {
+        Objects.requireNonNull(version);
+        this.version = version;
+        this.body = body;
+    }
+
+    @Override
+    public <T extends BinaryMessageVisitor<V>, V> V visit(BinaryMessageVisitorFactory<T, V> visitorFactory) throws MessageVisitException, IllegalStateException {
+        BinaryMessageVisitor<V> visitor = visitorFactory.createBinaryMessageVisitor(this.version);
+
+        // Grab from headers the attributes and extensions
+        this.forEachHeader((k, v) -> {
+            try {
+                if (k.substring(0, getHeaderKeyPrefix().length()).equalsIgnoreCase(getHeaderKeyPrefix())) {
+                    String name = k.substring(getHeaderKeyPrefix().length()).toLowerCase();
+                    if (name.equals("specversion")) {
+                        return;
+                    }
+                    if (this.version.getAllAttributes().contains(name)) {
+                        visitor.setAttribute(name, headerValueToString(v));
+                    } else {
+                        visitor.setExtension(name, headerValueToString(v));
+                    }
+                }
+                if (k.equalsIgnoreCase(getContentTypeHeaderKey())) {
+                    visitor.setAttribute("datacontenttype", headerValueToString(v));
+                }
+            } catch (StringIndexOutOfBoundsException ex) {
```

Comment:
You could also use `k.startsWith` to avoid this exception

Reply:
`startsWith` is not case insensitive

Reply:
then I'd suggest checking the length. Exceptions in Java are expensive.